File: Dijkstra.py
# link to question: https://codeforces.com/problemset/problem/20/C



import heapq

INF = float('inf')

n, m = map(int, input().split())
graph = [[] for _ in range(n + 1)]

for _ in range(m):
    u, v, w = map(int, input().split())
    graph[u].append((v, w))
    graph[v].append((u, w))

visited = [(INF, -1) for _ in range(n + 1)]  # (distance, previous_node)
visited[1] = (0, 0)

pq = []
heapq.heappush(pq, (0, 1))  # (distance, node)

while pq:
    dist, node = heapq.heappop(pq)
    if dist > visited[node][0]:
        continue
    for nxt, weight in graph[node]:
        new_dist = visited[node][0] + weight
        if new_dist < visited[nxt][0]:
            visited[nxt] = (new_dist, node)
            heapq.heappush(pq, (new_dist, nxt))

if visited[n][1] == -1:
    print(-1)
else:
    path = []
    node = n
    while node != 0:
        path.append(node)
        node = visited[node][1]
    print(' '.join(map(str, reversed(path))))

File: Dijkstra.cpp
// link to question: https://codeforces.com/problemset/problem/20/C



#include <bits/stdc++.h>
using namespace std;

#define all(a) (a).begin(), (a).end()
#define rall(a) (a).rbegin(), (a).rend()
#define sz(x) (int)(x).size()

#define rep(i, n) for (int i = 0; i < (n); ++i)
#define rep1(i, a, b) for (int i = (a); i <= (b); ++i)
#define per(i, n) for (int i = (n) - 1; i >= 0; --i)
#define per1(i, a, b) for (int i = (b); i >= (a); --i)

typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef vector<vector<ll>> vvll;
typedef vector<ll> vll;
typedef pair<int, int> pii;
typedef pair<ll, ll> pll;
using namespace std;

vector<vector<pair<int, int>>> graph;
vector<pair<ll, int>> visited;
ll check = LLONG_MAX;
int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    int n, m;
    cin >> n >> m;
    graph.resize(n + 1);
    rep(i, m)
    {
        int temp1, temp2, temp3;
        cin >> temp1 >> temp2 >> temp3;
        graph[temp2].push_back({temp1, temp3});
        graph[temp1].push_back({temp2, temp3});
    }
    visited.resize(n + 1, {check, -1});
    visited[1].first = 0;
    visited[1].second = 0;
    priority_queue<pair<ll, int>> pri;
    pri.push({0, 1});
    while (!pri.empty())
    {
        ll temp = pri.top().second;
        pri.pop();
        for (auto i : graph[temp])
        {
            if (visited[temp].first + i.second < visited[i.first].first)
            {
                visited[i.first].first = visited[temp].first + i.second;
                visited[i.first].second = temp;
                pri.push({-visited[i.first].first, i.first});
            }
        }
    }
    if (visited[n].second == -1)
    {
        cout << -1;
    }
    else
    {
        vi hi;
        int node = n;
        while (visited[node].second != -1)
        {
            hi.push_back(node);
            node = visited[node].second;
        }
        per(i, hi.size())
        {
            cout << hi[i] << ' ';
        }
    }
}
